airdrop/airdrop.gno

```diff
@@ -0,0 +1,215 @@
+	token         grc20.IGRC20
+	config        Config
+	claimedBitmap map[uint64]*u256.Uint
+	merkleTree    *merkle.Tree
+}
+
+type IAirdrop interface {
+	GetToken() string
+	GetConfig() Config
+	Claim(claim Claim, proof [][]byte) bool
+	Claim128(claims []Claim, remainingProof [][]byte) uint8
+	IsClaimed(claimID uint64) bool
+	Refund()
+}
+
+// NewAirdrop creates a new Airdrop instance
+func NewAirdrop(token grc20.IGRC20, config Config, claims []Claim) (*Airdrop, error) {
+	// create Merkle tree
+	//
+	// need to know the number of claims to create the hashables slice
+	hashables := make([]merkle.Hashable, len(claims))
+	for i, claim := range claims {
+		hashables[i] = claim
+	}
+	tree := merkle.NewTree(hashables)
+
+	// verify that the computed root matches the provided root
+	// if hex.EncodeToString(tree.Root()) != hex.EncodeToString(config.Root[:]) {
+	// 	return nil, errors.New("computed root does not match provided root")
+	// }
+
+	return &Airdrop{
+		token:         token,
+		config:        config,
+		claimedBitmap: make(map[uint64]*u256.Uint),
+		merkleTree:    tree,
+	}, nil
+}
+
+// GetToken returns the token address
+// func (a *Airdrop) GetToken() std.Address {
+// 	// TODO: get its address
+// 	return a.token.GetAddress()
+// }
+
+// GetConfig returns the airdrop configuration
+func (a *Airdrop) GetConfig() Config {
+	return a.config
+}
+
+// IsClaimed checks if a claim has been made (claimed)
+func (a *Airdrop) IsClaimed(claimID uint64) bool {
+	word, index := a.claimIDToBitmapIndex(claimID)
+	bitmap, exists := a.claimedBitmap[word]
+	if !exists {
+		return false
+	}
+	mask := new(u256.Uint).Lsh(u256.NewUint(1), uint(index))
+	return new(u256.Uint).And(bitmap, mask).Cmp(u256.Zero()) != 0
+}
+
+// claimIDToBitmapIndex converts a claim ID to bitmap index
+func (a *Airdrop) claimIDToBitmapIndex(claimID uint64) (word uint64, index uint8) {
+	word = claimID / 256
+	index = uint8(claimID % 256)
+	return
+}
+
+// Claim processes a single claim
+func (a *Airdrop) Claim(claim Claim, proof [][]byte) (bool, error) {
+	valid, err := a.verifyProof(claim, proof)
+	if err != nil {
+		return false, err
+	}
+	if !valid {
+		return false, ErrInvalidProof
+	}
+
+	word, index := a.claimIDToBitmapIndex(claim.ID)
+	bitmap, exists := a.claimedBitmap[word]
+	if !exists {
+		bitmap = u256.Zero()
+	}
+	mask := new(u256.Uint).Lsh(u256.NewUint(1), uint(index))
+	alreadyClaimed := new(u256.Uint).And(bitmap, mask).Cmp(u256.Zero()) != 0
+
+	if alreadyClaimed {
+		return false, nil // XXX should we return an error?
+	}
+
+	// update the bitmap
+	newBitmap := new(u256.Uint).Or(bitmap, mask)
+	a.claimedBitmap[word] = newBitmap
+
+	// transfer tokens
+	err = a.token.Transfer(std.Address(claim.Claimee), claim.Amount.Uint64())
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// verifyProof checks if the merkle proof is valid
+func (a *Airdrop) verifyProof(claim Claim, proff [][]byte) (bool, error) {
+	proof, err := a.merkleTree.Proof(claim)
```

Comment:
merkle.Verify(a.root, claim, proof) per comment above

Reply:
For now, this function has been removed as I have modified it to not use gno provided merkle trees (will find an alternative).